classes/Deck.js

```diff
@@ -1,22 +1,187 @@
-import { PlayingCard } from './PlayingCard'
+import { SUITES, RANKS } from '../constants/playing-card'
 
 class Deck {
-  constructor(suites, ranks) {
-    const cards = []
-    for (const suite of suites) {
-      for (const rank of ranks) {
-        cards.push(new PlayingCard(suite, rank))
+  constructor(cards = []) {
+    this.cards = cards
+  }
+
+  validateCanDrawCard(cardPosition) {
+    if ((0 >= cardPosition) || (cardPosition > this.size())) {
+      throw new Error(`RangeError: Cannot remove card from position (${cardPosition})`)
+    }
+  }
+
+  validateCanInsertCard(cardPosition) {
+    if ((0 >= cardPosition) || (cardPosition > this.size() + 1)) {
+      throw new Error(`RangeError: Cannot insert card to position ${cardPosition}`)
+    }
+  }
+
+  validateNotEmpty() {
+    if (this.isEmpty()) {
+      throw new Error(`IllegalOperation: Cannot draw card, deck is empty`)
+    }
+  }
+
+  validateNotFull() {
+    if (this.isFull()) {
+      throw new Error(`IllegalOperation: Cannot insert card, deck is full`)
+    }
+  }
+
+  print() {
+    this.cards.forEach(console.log)
+  }
+
+  drawCard(cardPosition) {
+    this.validateNotEmpty()
+    this.validateCanDrawCard(cardPosition)
+
+    return this.cards.splice(cardPosition - 1, 1)[0]
+  }
+
+  drawTopCard() {
+    return this.drawCard(this.size())
+  }
+  
+  drawBottomCard() {
+    return this.drawCard(1)
+  }
+
+  insertCard(card, cardPosition) {
+    this.validateNotFull()
+    this.validateCanInsertCard(cardPosition)
+
+    const { bottom, top } = this.split(cardPosition - 1)
+    bottom.cards.push(card)
+    
+    this.stack(bottom.stack(top))
+  }
+
+  insertCardTop(card) {
+    this.insertCard(card, this.size() + 1)
+  }
+
+  insertCardBottom(card) {
+    this.insertCard(card, 1)
+  }
+
+  isEmpty() {
+    return this.size() === 0
+  }
+
+  isFull() {
+    throw new Error(`Missing implementation for this function`)
+  }
+
+  size() {
+    return this.cards.length
+  }
+
+  stack(deck) {
+    this.cards = this.cards.concat(deck.cards)
+
+    return this
+  }
+
+  split(position) {
+    return { 
+      bottom: new Deck(this.cards.splice(0, position)),
+      top: new Deck(this.cards.splice(0, this.cards.length))
+    }
+  }
+
+  cut() {
+    const plusMinus = (Math.random() > 0.5) ? 1 : -1
+    const cut = Math.floor(Math.random() * Math.round(this.cards.length / 10)) * plusMinus
+
+    const { bottom, top } = this.split((this.cards.length / 2) + cut)
```

Comment:
1. You can use this.size()
2. For Decks with an odd number of cards, a non-integer is being passed in. That might not be a problem, but just conceptually it doesn't seem right to split the deck at a non-integer position